specs/data_structures.md

```diff
@@ -320,31 +286,178 @@ A proof into an SMT is structured as:
+
+For non-parity shares, if there is insufficient request data to fill the share, the remaining bytes are padded with `0`.
+
+## Arranging Available Data Into Shares
+
+The previous sections described how some original data, arranged into a `k * k` matrix, can be extended and committed to with NMT roots. This section specifies how [available data](#available-data) (which includes [transactions](#transactiondata), [intermediate state roots](#intermediatestaterootdata), [evidence](#evidencedata), and [messages](#messagedata)) is arranged into the matrix in the first place.
+
+ First, for each of `transactionData`, `intermediateStateRootData`, and `evidenceData`, [serialize](#serialization) the data and split it up into `SHARE_SIZE-SHARE_RESERVED_BYTES`-byte [shares](#share). This data has a _reserved_ namespace ID, and as such the first `SHARE_RESERVED_BYTES` bytes for these shares has special meaning. Then, concatenate the lists of shares in the order: transactions, intermediate state roots, evidence. Note that by construction, each share only has a single namespace, and that the list of concatenated shares is [lexicographically ordered by namespace ID](consensus.md#reserved-namespace-ids).
+
+These shares are arranged in the [first quadrant](#2d-reed-solomon-encoding-scheme) (`Q0`) of the `AVAILABLE_DATA_MATRIX_HALFSIZE*2 * AVAILABLE_DATA_MATRIX_HALFSIZE*2` available data matrix in _row-major_ order. In the example below, each reserved data element takes up exactly one share.
+
+![fig: Original data: reserved.](figures/rs2d_originaldata_reserved.svg)
+
+Each message in the list `messageData` is independently serialized and split into `SHARE_SIZE` shares. Then each message is padded by appending zero-shares (i.e. `SHARE_SIZE` bytes of `0x00`) until the smallest enclosing power of 2 (e.g. a message that fits into three shares is padded to four shares). For each message, the following algorithm is used to place it in the available data matrix, with row-major order:
+1. Place the first share of the message at the next unused location in the matrix whose column in aligned with the number of padded shares (i.e. if there are four padded shares, then only every fourth location can be used to start) **unless** there are insufficient unused locations in the row. Then place the remaining shares in the following locations.
+1. If there are insufficient unused locations in the row, place the first share of the message at the first column of the next row. Then place the remaining shares in the following locations.
+
+In the example below, two messages (of padded sizes 2 and 1) are placed following the aforementioned rules.
+
+![fig: Original data: messages.](figures/rs2d_originaldata_message.svg)
+
+By construction, this gives a useful property: transactions [can commit to a commitment to a list of roots](#transaction) that are each guaranteed (assuming the block is valid) to be subtree roots in one or more of the row NMTs.
```

Comment:
If `k` is a power of 2, then using the two rules guarantees (unless I'm mistaken) that the transaction sender knows enough about the layout of the message in a valid block to know how it will be split up across rows (if it is) and thus how to commit to it.

As an example, suppose `k = 8` and the number of shares of the message is `16`. The transaction sender knows it'll take up two rows (they don't know which two exactly, just some two). Since they commit to row subtree roots, they can just commit to the roots of those two rows, then the root of those roots.

To verify without downloading all the message data, we know that a message of size `16` takes up two rows. So we can ask for both subtree roots that are behind the hash that's in the transaction, and check that they're included at the rows indicated by `messageStart` that the block producer sets.